### PR TITLE
Enable Enum Support in MoccaClient

### PR DESCRIPTION
**Pull request by sharathkuncham**
Enable support for Enum as parameter, return type and complex type in GraphQL Mocca Client.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -9,7 +9,7 @@ Set testModules = [
 subprojects {
 
     group = 'com.paypal.mocca'
-    version = '0.0.8-SNAPSHOT'
+    version = '0.0.8'
```
Comment (fabiocarvalho777) [inline build.gradle]: Please, undo this change. This will be done in a separated PR.

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java [hunk 2]
```diff
@@ -171,6 +171,9 @@ private static void configureMutationStubs() {
         final String EXPECTED_NULL_PARAMETER_REQUEST = "{\"query\":\"mutation{addSample(foo: \\\"zoo\\\", bar: \\\"\\\") {bar foo}}\"}";
         final String NULL_PARAMETER_RESULT = "{\"data\": {\"addSample\": {\"foo\": \"zoo\",\"bar\": \"\"}}}";
 
+        final String EXPECTED_ENUM_REQUEST = "{\"query\":\"query{addEnum(\\\"Sample1\\\")}\"}";
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]: Why did you remove the parameter name from here? How is the client supposed to know which parameter is which if there are multiple ones? 
This line was not supposed to be changed, only the response one.

Please add it back.

File: mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java [hunk 3]
```diff
@@ -150,6 +150,13 @@ public void queryUUIDTest() {
         assertEquals(actualUUID, expectedUUID);
     }
 
+    @Test
+    public void queryEnumTest() {
+        SampleEnum expectedSampleEnum = SampleEnum.Sample1;
+        SampleEnum actualSampleEnum = client.addEnum(SampleEnum.Sample1);
+        assertEquals(actualSampleEnum,expectedSampleEnum);
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]:
```suggestion
        assertEquals(actualSampleEnum, expectedSampleEnum);
```

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 4]
```diff
@@ -233,15 +235,25 @@ private static boolean isPojo(Type type) {
         return !clazz.isPrimitive();
     }
 
+
+    private static boolean isEnum(Type type) {
+        return  Enum.class.isAssignableFrom((Class<?>) type);
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]:
```suggestion
        return Enum.class.isAssignableFrom((Class<?>) type);
```

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 5]
```diff
@@ -448,6 +460,11 @@ private void writeResponseTypeSelectionSet(final ByteArrayOutputStream requestPa
             final Type listResponseType = getInnerType(cfResponseType, List.class).orElse(cfResponseType);
             final Type rawResponseType = getInnerType(listResponseType, Optional.class).orElse(listResponseType);
 
+            if(isEnum(rawResponseType)){
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]:
```suggestion
            if (isEnum(rawResponseType)) {
```

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 6]
```diff
@@ -233,15 +235,25 @@ private static boolean isPojo(Type type) {
...
-                prefix + "\\\"" + value.toString() + "\\\"" :
-                prefix + value.toString();
+        if(Enum.class.isAssignableFrom((Class<?>) type)) {
+            return "\\\"" + value + "\\\"";
+        } else {
+            return type == String.class || type == Character.class || type == OffsetDateTime.class || type == Duration.class
+                    || type == UUID.class || type.getTypeName().equals("char") ?
+                    prefix + "\\\"" + value.toString() + "\\\"" :
+                    prefix + value.toString();
+        }
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]:
```suggestion
        return type == String.class || type == Character.class || type == OffsetDateTime.class || type == Duration.class
                || type == UUID.class || type.getTypeName().equals("char") || Enum.class.isAssignableFrom((Class<?>) type) ?
                prefix + "\\\"" + value.toString() + "\\\"" :
                prefix + value.toString();	
```

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 6 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java:256]: I didn't get this one. Aren't we supposed to remove the prefix if its Enum as that was the concern.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 5 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: In my IDE the spacing is right. Not sure why its not showing here in PR.

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java [hunk 7]
```diff
@@ -171,6 +171,9 @@ private static void configureMutationStubs() {
         final String EXPECTED_NULL_PARAMETER_REQUEST = "{\"query\":\"mutation{addSample(foo: \\\"zoo\\\", bar: \\\"\\\") {bar foo}}\"}";
         final String NULL_PARAMETER_RESULT = "{\"data\": {\"addSample\": {\"foo\": \"zoo\",\"bar\": \"\"}}}";
 
+        final String EXPECTED_ENUM_REQUEST = "{\"query\":\"query{addEnum(sampleEnum: \\\"Sample1\\\")}\"}";
+        final String ACTUAL_ENUM_RESPONSE = "{\"data\": {\"addEnum\": \"Sample1\"}}";
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]:
```suggestion
        final String ENUM_RESPONSE = "{\"data\": {\"addEnum\": \"Sample1\"}}";
```

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java [hunk 8]
```diff
@@ -180,6 +183,7 @@ private static void configureMutationStubs() {
         addGraphQlStub(EXPECTED_NULL_PARAMETER_REQUEST, NULL_PARAMETER_RESULT, DEFAULT_HEADERS);
         addGraphQlStub(EXPECTED_ERROR_REQUEST, ERROR_RESULT, DEFAULT_HEADERS);
         addGraphQlStub(EXPECTED_ERROR_LIST_REQUEST, ERROR_LIST_RESULT, DEFAULT_HEADERS);
+        addGraphQlStub(EXPECTED_ENUM_REQUEST,ACTUAL_ENUM_RESPONSE,DEFAULT_HEADERS);
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]:
```suggestion
        addGraphQlStub(EXPECTED_ENUM_REQUEST, ENUM_RESPONSE, DEFAULT_HEADERS);
```

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java (hunk 8 above)
Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]: Ack

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java (hunk 7 above)
Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]: Ack

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java (hunk 8 above)
Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]: Done

File: mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java (hunk 7 above)
Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/WireMockProvider.java]: Done

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 9]
```diff
@@ -246,10 +246,14 @@ private static boolean isEnum(Type type) {
      */
     private String writeRequestVariable(String name, Object value, Type type) {
         final String prefix = name == null ? "" : name + ": ";
-        return type == String.class || type == Character.class || type == OffsetDateTime.class || type == Duration.class
-                || type == UUID.class || type.getTypeName().equals("char") || Enum.class.isAssignableFrom((Class<?>) type) ?
-                prefix + "\\\"" + value.toString() + "\\\"" :
-                prefix + value.toString();
+        if(Enum.class.isAssignableFrom((Class<?>) type)) {
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Can we use `isEnum` here?

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 9 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Yes. Missed it. Fixed now.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 10]
```diff
@@ -187,7 +187,9 @@ private void writeRequestVariables(final ByteArrayOutputStream requestPayload, f
                         value = variable.value;
                     }
 
-                    if (isPojo(type)) {
+                    if(isEnum(type)) {
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Undo this change please. it is not necessary.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 11]
```diff
@@ -233,15 +235,25 @@ private static boolean isPojo(Type type) {
...
      * Returns a String containing the specification of a GraphQL variable name and its value,
      * as it is supposed to be written in the request payload
      */
     private String writeRequestVariable(String name, Object value, Type type) {
         final String prefix = name == null ? "" : name + ": ";
-        return type == String.class || type == Character.class || type == OffsetDateTime.class || type == Duration.class || type == UUID.class || type.getTypeName().equals("char") ?
-                prefix + "\\\"" + value.toString() + "\\\"" :
-                prefix + value.toString();
+        if(isEnum(type)) {
+           return prefix + value.toString();
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Why would we not print the variable name for **request** parameter enumerations?

Notice this method is specific to writing **request variables**, and enumerations parameters must have the parameter name. The **return** value is the one that doesn't need it.

I believe this change here is wrong, it should be like you did before, with an additional condition (for enum) added to the end of line 253.

I understand the tests, and GRE, are functional regardless of what I am saying in this comment. However, that to me just means there is another bug somewhere that is making this change functional in spite of that.

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 5 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Fixed the formatting.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 11 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Here prefix is the one which represents the parameter name. So parameter name is in there. The reason I have to separate it out is because we want to get rid of quotes for enum value as the graphql layer does not like the enum value with quotes around it and it triggers the exception. It's similar to line 253 just that it's taking statement in else condition of the ternary operator, the variable name plus variable value without quotes.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 10 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Screen Shot 2023-11-15 at 5 49 26 PM
Without this check Enum will satisfy the isPojo condition(type instanceof class) and hence it will follow the code path for POJO. To avoid this we need to predetermine if its Enum type first.

Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: I see. In that case, please, what if we add back for enum under `NON_POJO_TYPES`? If that does not work, then please add a check for enum under `isPojo` method.

So, let's go ahead and remove this condition from here please.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 11 above)
Comment (fabiocarvalho777) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Understood, it makes sense. Thanks for the clarification.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 10 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Adding Enum under NON_POJO_TYPES wont work because the check at line number 229 (NON_POJO_TYPES.contains(type))) would fail for Enum type. contains wont work for actual Enum type(ex - SampleEnum). I will try adding isEnum check under isPojo method and check. Thanks!

Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Done. Removed the condition and added enum check under isPojo method.

File: mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java [hunk 12]
```diff
@@ -194,22 +194,24 @@ public void queryComplexDataTest() {
     }
 
     @Test
-    public void queryOptionalTest() {
-
+    public void queryOptionalTestDTO(){
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]:
```suggestion
    public void queryOptionalSimpleTest() {
```

File: mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java [hunk 13]
```diff
@@ -194,22 +194,24 @@ public void queryComplexDataTest() {
...
         SampleRequestDTO sampleRequestDTO = new SampleRequestDTO("boom", "zaz");
         Optional<SampleResponseDTO> result = client.getOneSample(Optional.of(sampleRequestDTO));
         assertNotNull(result);
         assertTrue(result.isPresent());
         assertEquals(result.get().getFoo(), "boo");
         assertEquals(result.get().getBar(), "far");
+    }
 
+    @Test
+    public void queryOptionalTest() {
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]:
```suggestion
    public void queryOptionalComplexTest() {
```

File: mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java [hunk 14]
```diff
@@ -223,7 +225,7 @@ public void queryOptionalTest() {
         assertTrue(superComplexResponse.getOptionalField().isPresent());
         assertEquals(superComplexResponse.getOptionalField().get(), "love");
         assertEquals(superComplexResponse.getDuration(), Duration.ofHours(3));
-        assertEquals(superComplexResponse.getUuid(), UUID.fromString("229c07ba-04bc-49a6-13bc-165e1a54cb33"));
+        assertEquals(superComplexResponse.getUuid(), UUID.fromString("229c07ba-04bc-49a6-13bc-165e1a54cb32"));
```
Comment (fabiocarvalho777) [inline mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]: Out of curiosity, why did this change?

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java (hunk 14 above)
Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]: As am debugging the stub mismatch I was trying few things. One is this to differentiate it from other complex test. Just for my help. I will roll it back to previous state once I figure out the stub issue.

Comment (sharathkuncham) [reply mocca-client/src/test/java/com/paypal/mocca/client/MoccaClientQueryTest.java]: Reverted the value.

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java [hunk 15]
```diff
@@ -355,7 +355,9 @@ private String objectToString(final Object object, final String name, final List
         } else if (object instanceof Optional) {
             Optional<?> optional = (Optional) object;
             return optional.map(v -> objectToString(v, name, ignoreFields)).orElse("null");
-        } else {
+        } else if (object instanceof Enum) {
```
Comment (fabiocarvalho777) [inline mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Undo this change please, and add an enum condition in line 345

**Review by fabiocarvalho777 — CHANGES_REQUESTED** [review]

File: mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java (hunk 15 above)
Comment (sharathkuncham) [reply mocca-client/src/main/java/com/paypal/mocca/client/MoccaSerializer.java]: Done

**Review by fabiocarvalho777 — APPROVED** [review]
